### DISCUSSION TITLE
Console warning: `API handler should not return a value`

### DISCUSSION THREAD

**Discussion by Sheraff** [vercel/next.js#48951] [Help] [answered]
### Summary

I've recently started receiving `API handler should not return a value, received object.` errors, which (thanks to GitHub's amazing "search everything everywhere all at once" feature) I was able to trace back to this bit of code in next.js

https://github.com/vercel/next.js/blob/20d8e93d1e95015621b2cb095e26210af1e8a263/packages/next/src/server/api-utils/node.ts#L590-L600

I've recently upgraded from `"next": "13.1.4",` to `"next": "13.3.1",` so I'm assuming this log is recent-ish (or at least it's the first time I've noticed it).

Everything I return from API routes is of the TS type `NextApiResponse` or `undefined` as far as I can tell:
```ts
return res.status(500).json({ error: "Error transforming image" })
//                     ^? (body: any) => void
```
```ts
return returnStream.pipe(res)
//                  ^? (res: NextApiResponse) => NextApiResponse
```
```ts
return res.end()
//         ^? internal.Writable.end(cb?: (() => void) | undefined): NextApiResponse
```
unless it comes from a 3rd party route like [tRPC](https://trpc.io/) (`createNextApiHandler()`) or [next-auth](https://next-auth.js.org/) (`NextAuth()`).

- What should in fact be returned from an API route, if anything? 
- Did something change in the specs or was it always an error and just recently started being logged to the console?
- Could the log itself be a little more explicitly saying that it "comes from Next"? (because if it wasn't for the GitHub search feature, I would have had a hard time tracing it back to Next)

**Comment by sunech** [comment]
I started getting the same message, after upgrading from 13.2.3 to 13.4.1.
Haven't identified the cause yet, but I'm not using tRPC or next-auth.

**Reply by Sheraff** [reply]
> I started getting the same message, after upgrading from 13.2.3 to 13.4.1. Haven't identified the cause yet, but I'm not using tRPC or [next-auth](https://linear.app/vercel/issue/NEXT-auth).

I actually figured it out. Basically you're not supposed to return anything from your routes, but you can return `res.end()` for example (or whatever else that has the same signature).


Also, my `return returnStream.pipe(res)` call wasn't correctly typed, and it does not in fact return a `NextApiResponse`, so this was the culprit.

**Reply by sunech** [reply]
Thanks for sharing - I think I have a lot of API pages that needs updating then 🙈

**Reply by heymartinadams** [reply]
Wait what!? @Sheraff:

> I actually figured it out. Basically you're not supposed to return anything from your routes

Official NextJS examples explicitly [show a return](https://vercel.com/docs/concepts/functions/edge-functions#regional-edge-function-invocation):

```js
return NextResponse.json({  name: `Hello, from ${request.url} I'm now an Edge Function!` });
```

Nevertheless, running into the same issue. I do return `NextResponse.json()` on every edge function.

🤔

**Reply by Sheraff** [reply]
`.json()` has a signature of `(body: any) => void` so it's like returning nothing.

**Reply by pastilline33** [reply]
return res.status(204).json(); bingo

**Comment by ilhantekir** [comment]
The main problem is that we are not using env files correctly.

I don't understand the real purpose of this code snippet why does it give a warning if not production mode and not apiRouteResult undefined :)

I was seeing this warn in the error logs on production and this was the solution for me; I don't use env files in the root, i have a folder called env and a file called .env.production and that's how I use it;

`"scripts": {
    "build:Production": "NODE_ENV=production env-cmd -f env/.env.production next build",
    "start:Production": "NODE_ENV=production env-cmd -f env/.env.production next start",
     ...
  },`

there is no problem when we do the build through production flag (NODE_ENV=production) but it does not work if the build without using the production flag

for more information: [https://nextjs.org/docs/app/building-your-application/configuring/environment-variables](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables)

**Reply by wearon** [reply]
how does one identify which API route is the culprit?

**Reply by jt561** [reply]
Yeah, I need to figure this out too. I see the message but do not know where it's coming from!

**Reply by Sheraff** [reply]
The code responsible for this error is highlighted in the original question. It happens in a context that has access to the request and the query. You could
- edit the source code locally right inside your node_modules to add info and debug your situation
- submit a PR to Next to add this info for everyone

**Reply by jt561** [reply]
I think you misunderstood my question. What I’m looking for is the api route in **_my_** code that’s causing the error to be thrown. 

I can see the code that throws the error from NextJS’s side but from my side, there isn’t really an “error”, it just prints a warning message, so I have no clue which route is causing that message to be printed

**Reply by sunech** [reply]
Not sure if you can identify it, but instead you can try the approach of searching the entire api folder for "return " statements and start replacing them all. I did it this way and got rid of the error.

**Reply by Sheraff** [reply]
> I have no clue which route is causing that message to be printed

Sorry my answer wasn't very clear, I was proposing that you add some clues to which route it is in the logs. Since the location of that log is in a scope that has access to `req` and `query`, maybe there are some properties of these objects that could be added to the log to make it easier to trace it back to the route responsible for the error.

**Reply by jt561** [reply]
Oh I see what you mean. Thanks 😊

**Comment by EWebD** [comment]
This helped me solved the warning. I just removed return statement from NextAuth wrapper for using req,res.

before:
export default async function auth(req: NextApiRequest, res: NextApiResponse) { return await NextAuth(req, res, authOptions); }

after:
export default async function auth(req: NextApiRequest, res: NextApiResponse) { await NextAuth(req, res, authOptions); }

**Reply by prichodko** [reply]
thanks, this fixed the issue for me

**Comment by lamiaagebriel** [comment]
I recently encountered a similar issue while using Next.js v13.4.12, Next-Auth v4.22.3, and Prisma v5.0.0.

The problem arose from a conflict between providers in the authentication setup. Upon investigating, I realized that I inadvertently copied code from one authentication provider (GitHub) and overlooked changing it to the appropriate provider (Google), which caused the authentication process to fail. Therefore, it is crucial to ensure that all authentication-related code aligns correctly with the chosen provider.

Additionally, I had defined routes in both the 'pages' and 'app' folders, which may have contributed to intermittent functionality. To eliminate potential sources of error, it's advisable to focus on one file and verify if the issue persists.

Please ensure that you have accurately configured the authentication provider and review the routes setup to resolve the problem effectively.

**Comment by jonathanmv** [comment]
For people using `next-auth` and having this warning, make sure to update your `next-auth` version. The issue was fixed by [this PR](https://github.com/nextauthjs/next-auth/issues/8234)

**Comment by GorvGoyl** [comment]
I figured it out for serverless functions: don't return the response object. just use it like below:

```js
res.status(200).send({ data: mydata});
```

**Reply by Tigatok** [reply]
may be a dumb question, does `send` kill the line? Or would you have to wrap something in a if/else? Example

```
If(something){
  res.status(200)
}
res.status(500)
```

Or just add `return` below?

**Reply by jt561** [reply]
@Tigatok 
I think send returns a response to the api caller but continues running the rest of the code on your server so you would need to return like this:

```javascript
if (true) {
    return res.status(200);
}

return res.status(500);
```

or even 

```javascript
if (true) {
    res.status(200);
    return;
}

res.status(500);
return;
```

I can’t remember if setting the status is enough but you might have to have `.send` after setting the status to actually send it.

**Reply by jt561** [reply]
@Tigatok 
Also, you can add syntax highliting to your code snippets by specifying the language after 3 backticks so it would look like this in plain text:

‘’’javascript
if (true) { 
    console.log(“ha”);
}
’’’

I’ve used single quotes here so it doesn’t turn it into actual code.

**Comment by Lois-Varghese** [comment]
Any update on this? I'm using next with GraphQL, Apollo server, client and SWR in my react project. I'm not using the NextAuth and still get this error even though my api requests return res.status(200).send(data)

**Reply by Sheraff** [reply]
Basically: you don't need to return anything from API routes.

**Reply by Lois-Varghese** [reply]
Fixed my issue, I just had to update the appollo-server-integration-next package, it wasn't supporting route handlers. Had to go through tons of docs just to get here https://github.com/apollo-server-integrations/apollo-server-integration-next/pull/84 and understand the issue.

**Comment by Sheraff** [comment, marked as answer]
For anyone stumbling here: the answer is **only return `undefined` from your API routes**.

**Reply by idealistraj9** [reply]
ok

**Reply by mausworks** [reply]
If you want to do a one-liner, like so:

```ts
if (something) return res.status(400).end();
```

Just use the `void` operator:

```ts
if (something) return void res.status(400).end();
```

From [the MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void) for the void operator:

> The `void` operator evaluates the given expression and then returns `undefined`.

**Comment by DmitryKandratsenka** [comment]
@Sheraff hello, I'm kind of disagree with this console warning. Why can't I return anything I want if it doesn't affect anything? For example, I'm using the library `next-http-proxy-middleware` and my call looks like that:
```
  return httpProxyMiddleware(req, res, options);
```

This function `httpProxyMiddleware` returns some object, that is not used by `next` - that's fine. I can replace it by
```
  await httpProxyMiddleware(req, res, options);
  return;
```

...but why should I do it because of this console warning? I prefer this syntax in my function: `return httpProxyMiddleware(req, res, options);`

Could you, please, remove this warning, since it doesn't affect anything and just annoying the people?

**Comment by boardy** [comment]
This may have been fixed in NextJs 14, not sure as can't use it, but on 13 where I see this error I have to send undefined in the json function to stop this error being outputted. Shouldn't need to do this but at least stops the noise of the warning

```
return res.status(204).json(undefined);
```

**Reply by mausworks** [reply]
Or, you can use the `void` operator, which I think is more to the point:

```ts
return void res.status(204).end();
```

From [the MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void) for the `void` operator:

> The `void` operator evaluates the given expression and then returns `undefined`.
